packages/slice-machine/server/src/api/push-changes.ts

```diff
@@ -0,0 +1,197 @@
+  type SliceInsertChange,
+  type SliceUpdateChange,
+  type Limit,
+  type SliceDeleteChange,
+  type CustomTypeInsertChange,
+  type CustomTypeUpdateChange,
+  type CustomTypeDeleteChange,
+  type BulkBody,
+  type ClientError,
+  ChangeTypes,
+} from "@slicemachine/client";
+import * as Libraries from "@slicemachine/core/build/libraries";
+import { Slices, type SliceSM } from "@slicemachine/core/build/models/Slice";
+import {
+  CustomTypes,
+  type CustomTypeSM,
+} from "@slicemachine/core/build/models/CustomType";
+import { getLocalCustomTypes } from "../../../lib/utils/customTypes";
+import {
+  ModelStatus,
+  computeModelStatus,
+} from "../../../lib/models/common/ModelStatus";
+import { onError } from "../../../lib/models/server/ApiResult";
+import type { ApiResult } from "../../../lib/models/server/ApiResult";
+import type { Component, Library } from "@slicemachine/core/build/models";
+import type {
+  LocalOrRemoteCustomType,
+  LocalOrRemoteSlice,
+} from "../../../lib/models/common/ModelData";
+
+import { normalizeFrontendSlices } from "../../../lib/models/common/normalizers/slices";
+import { normalizeFrontendCustomTypes } from "../../../lib/models/common/normalizers/customType";
+import { BackendEnvironment } from "../../../lib/models/common/Environment";
+import { PushChangesPayload } from "../../../lib/models/common/TransactionalPush";
+
+type TransactionalPushBody = {
+  body: PushChangesPayload;
+  env: BackendEnvironment;
+};
+
+export default async function handler({
+  body,
+  env,
+}: TransactionalPushBody): Promise<
+  ApiResult<{ status: number; body: Limit | null }>
+> {
+  const { cwd, client, manifest } = env;
+
+  if (!manifest.libraries)
+    return {
+      status: 400,
+      body: null,
+    };
+
+  /* -- Retrieve all the different models -- */
+  const remoteCustomTypes: CustomTypeSM[] = await client
+    .getCustomTypes()
+    .then((customTypes) =>
+      customTypes.map((customType) => CustomTypes.toSM(customType))
+    );
+
+  const remoteSlices: SliceSM[] = await client
+    .getSlices()
+    .then((slices) => slices.map((slice) => Slices.toSM(slice)));
+
+  const localCustomTypes: CustomTypeSM[] = getLocalCustomTypes(cwd);
+
+  const localSlices: ReadonlyArray<Library<Component>> = Libraries.libraries(
+    cwd,
+    manifest.libraries
+  );
+
+  /* -- Assemble the models together and compute their statuses -- */
+  const slicesModels: ReadonlyArray<LocalOrRemoteSlice> =
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    normalizeFrontendSlices(localSlices, remoteSlices);
+  const customTypeModels: ReadonlyArray<LocalOrRemoteCustomType> =
+    Object.values(
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      normalizeFrontendCustomTypes(localCustomTypes, remoteCustomTypes)
+    );
+
+  /* -- Compute the request body -- */
+
+  const sliceChanges = slicesModels.reduce(
+    (
+      acc: (SliceInsertChange | SliceUpdateChange | SliceDeleteChange)[],
```

Comment:
I would have thought so too 😓  but i think it's because we change the original type that typescript doesn't like if we don't declare the type of the accumulator.